### PR TITLE
feat: cache resolved git/helm repository revisions

### LINKED ISSUE #1399: Option to disable polling
**Is your feature request related to a problem? Please describe.**
We would like the option to disable git polling completely and as we only want to use the web hook for triggering updates.

**Describe the solution you'd like**
Option to disable git polling.

**Describe alternatives you've considered**
I know we can increase the polling interval, but it would be nice to just have the option to disable it completely.

### PR DESCRIPTION
**Pull request by alexmt**
Signed-off-by: Alexander Matyushentsev <AMatyushentsev@gmail.com>

Closes #1399

PR adds caching of `git ls-remote`, and helm index and significantly reduces the number of requests to Git/Helm repository.  It is especially effective for mono-repos with multiple argocd apps. Example below is Argo CD that has 600 apps in 4 repos:

image

The revision information is cached by repo server. The cache duration must much the application reconciliation timeout. Instead of asking user to set the same timeout in two places PR promotes `--app-resync` flag to `timeout.reconciliation` key in `argocd-cm` ConfigMap that is bound to ENV variable in both application controller and repo server deployments. The change is backward compatible. It is described in upgrade instructions.

### CODE REVIEW

File: docs/operator-manual/upgrading/2.0-2.1.md [hunk 1]
```diff
@@ -5,4 +5,11 @@
...
-From here on you can follow the [regular upgrade process](./overview.md).
\ No newline at end of file
+From here on you can follow the [regular upgrade process](./overview.md).
+
+## Replacing `--app-resync` flag with `timeout.reconciliation` setting
+
+The`--app-resync` flag allows controlling how frequently Argo CD application controller checks resolve the target
+application revision of each application. In order to allow caching resolved revision per repository as opposed to per
+application, the `--app-resync` flag has been deprecated. Please use `timeout.reconciliation` setting instead.
+See example in [argocd-cm.yaml](../argocd-cm.yaml).
```
Comment (jannfis) [inline docs/operator-manual/upgrading/2.0-2.1.md:15]: > Please use `timeout.reconciliation` setting instead.

Should be precised as:

> Please use `timeout.reconciliation` setting in `argocd-cm` ConfigMap instead.

File: reposerver/cache/cache.go [hunk 2]
```diff
@@ -69,6 +74,43 @@ func (c *Cache) SetApps(repoUrl, revision string, apps map[string]string) error
...
+	var input [][2]string
+	if err := c.cache.GetItem(gitRefsKey(repo), &input); err != nil {
+		return err
+	}
+	var res []*plumbing.Reference
+	for i := range input {
+		res = append(res, plumbing.NewReferenceFromStrings(input[i][0], input[i][1]))
+	}
+	*references = res
+	return nil
```
Comment (jannfis) [inline reposerver/cache/cache.go:111]: Can you please add some comments to the public methods?

File: server/server.go [hunk 3]
```diff
@@ -726,7 +726,7 @@ func (a *ArgoCDServer) newHTTPServer(ctx context.Context, port int, grpcWebHandl
 	a.registerDexHandlers(mux)
 
 	// Webhook handler for git events
-	acdWebhookHandler := webhook.NewHandler(a.Namespace, a.AppClientset, a.settings, a.settingsMgr, repocache.NewCache(a.Cache.GetCache(), 24*time.Hour))
+	acdWebhookHandler := webhook.NewHandler(a.Namespace, a.AppClientset, a.settings, a.settingsMgr, repocache.NewCache(a.Cache.GetCache(), 24*time.Hour, 3*time.Minute))
```
Comment (jannfis) [inline server/server.go:729]: I was wondering why we use hard-coded values for cache expiration here

File: util/git/client.go [hunk 4]
```diff
@@ -87,23 +103,41 @@ func init() {
...
+		c.gitRefCache = cache
+		c.loadRefFromCache = loadRefFromCache
+	}
+}
+
+func WithEventHandlers(handlers EventHandlers) ClientOpts {
+	return func(c *nativeGitClient) {
+		c.EventHandlers = handlers
+	}
+}
```
Comment (jannfis) [inline util/git/client.go:119]: Can we please have some comments on public methods here?

File: docs/operator-manual/upgrading/2.0-2.1.md [hunk 5]
```diff
@@ -5,4 +5,11 @@
 Note that bundled Kustomize has been upgraded to v4.1.2. Some of the flags are changed in Kustomize V4. 
 For example flag name `load_restrictor` is changed in Kustomize v4+. It is changed from `--load_restrictor=none` to `--load-restrictor LoadRestrictionsNone`. 
 
-From here on you can follow the [regular upgrade process](./overview.md).
\ No newline at end of file
+From here on you can follow the [regular upgrade process](./overview.md).
```
Comment (jannfis) [inline docs/operator-manual/upgrading/2.0-2.1.md:8]: This should be moved to the bottom.

**Review by jannfis — CHANGES_REQUESTED** [review]
Great change, @alexmt - reducing the load on external systems is always important.

I have some minor comments, see below.

File: cmd/argocd-application-controller/commands/argocd_application_controller.go [hunk 6]
```diff
@@ -137,7 +137,7 @@ func NewCommand() *cobra.Command {
 	}
 
 	clientConfig = cli.AddKubectlFlagsToCmd(&command)
-	command.Flags().Int64Var(&appResyncPeriod, "app-resync", defaultAppResyncPeriod, "Time period in seconds for application resync.")
+	command.Flags().Int64Var(&appResyncPeriod, "app-resync", int64(env.ParseDurationFromEnv("RECONCILIATION_TIMEOUT", defaultAppResyncPeriod, 0, math.MaxInt32).Seconds()), "Time period in seconds for application resync.")
```
[the remark this answers is not in the document]
Comment (alexmt) [reply cmd/argocd-application-controller/commands/argocd_application_controller.go:140]: 1. Totally agree - renamed.
2. Done
3. If someone specifies "3mins" then `.Seconds()` should return 180. This way I can avoid changing `--app-resync` flag and distracting users who already rely on it.

File: docs/operator-manual/upgrading/2.0-2.1.md (hunk 1 above)
Comment (alexmt) [reply docs/operator-manual/upgrading/2.0-2.1.md:15]: done

File: reposerver/cache/cache.go (hunk 2 above)
Comment (alexmt) [reply reposerver/cache/cache.go:111]: done

File: server/server.go (hunk 3 above)
Comment (alexmt) [reply server/server.go:729]: API server does not write to cache and does not really use these values. Added comment to clarify it. Do you think it is good enough?

File: util/git/client.go (hunk 4 above)
Comment (alexmt) [reply util/git/client.go:119]: done

File: docs/operator-manual/upgrading/2.0-2.1.md (hunk 5 above)
Comment (alexmt) [reply docs/operator-manual/upgrading/2.0-2.1.md:8]: thank you. done.

**Comment by alexmt** [comment]
Thank you for review @jannfis . PTAL

**Review by jannfis — APPROVED** [review]
Thanks @alexmt 

I found a few spots where `RECONCILIATION_TIMEOUT` was not renamed to `ARGOCD_RECONCILIATION_TIMOUT`, please change those too before merging.

Otherwise, LGTM!

**Comment by alexmt** [comment]
Oh, thank you! Totally forgot to replace `RECONCILIATION_TIMEOUT` everywhere. It is fixed now.

**Pull request merged**